### PR TITLE
IP generic validator.

### PR DESCRIPTION
**Pull request by gguridi**
Hello!

* Type: new feature
* Link to issue:

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change:
- This is a generic ip validator for general purposes. The MVC models have one but it couldn't be used for generic forms so I've created if with the unit tests associated.

Thanks

---
This change is [Reviewable](https://reviewable.io/reviews/phalcon/cphalcon/12513)

### CODE REVIEW

**Comment by Jurigag** [comment]
Reopen PR cuz failing tests seems unrelated.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 - Added afterBinding event to `Phalcon\Dispatcher` and `Phalcon\Mvc\Micro`, added `Phalcon\Mvc\Micro::afterBinding`
 - Added the ability to set custom Resultset class returned by find() [#12166](https://github.com/phalcon/cphalcon/issues/12166)
 - Added the ability to clear appended and prepended title elements (Phalcon\Tag::appendTitle, Phalcon\Tag::prependTitle). Now you can use array to add multiple titles. For more details check [#12238](https://github.com/phalcon/cphalcon/issues/12238).
+- Added ip generic validator so we can add ip textfields into forms, etc.
```
Comment (Jurigag) [inline CHANGELOG.md]: I would like more:
> Added `Phalcon\Validation\Validator\Ip` ... IP ...

**Comment by gguridi** [comment]
Added the changes requested in the changelog file.

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@
 - Added afterBinding event to `Phalcon\Dispatcher` and `Phalcon\Mvc\Micro`, added `Phalcon\Mvc\Micro::afterBinding`
 - Added the ability to set custom Resultset class returned by find() [#12166](https://github.com/phalcon/cphalcon/issues/12166)
 - Added the ability to clear appended and prepended title elements (Phalcon\Tag::appendTitle, Phalcon\Tag::prependTitle). Now you can use array to add multiple titles. For more details check [#12238](https://github.com/phalcon/cphalcon/issues/12238).
+- Added `Phalcon\Validation\Validator\Ip`, class used to validate ip address fields. It allows to validate a field selecting IPv4 or IPv4, allowing private or reserved ranges and empty values if necessary.
```
Comment (Jurigag) [inline CHANGELOG.md]: IPv4 OR IPv**6**

Comment (gguridi) [inline CHANGELOG.md]: Damn! :) Fixed.

File: phalcon/validation/validator/ip.zep [hunk 3]
```diff
@@ -0,0 +1,148 @@
+
+/*
+ +------------------------------------------------------------------------+
+ | Phalcon Framework                                                      |
+ +------------------------------------------------------------------------+
+ | Copyright (c) 2011-2016 Phalcon Team (https://phalconphp.com)       |
```
Comment (sergeyklay) [inline phalcon/validation/validator/ip.zep]: 1. `2011-2017`
2. Could you please fix the indent issue

File: phalcon/validation/validator/ip.zep [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ +------------------------------------------------------------------------+
+ | This source file is subject to the New BSD License that is bundled     |
+ | with this package in the file docs/LICENSE.txt.                        |
+ |                                                                        |
+ | If you did not receive a copy of the license and are unable to         |
+ | obtain it through the world-wide-web, please send an email             |
+ | to license@phalconphp.com so we can send you a copy immediately.       |
+ +------------------------------------------------------------------------+
+ | Authors: Gorka Guridi <gorka.guridi@gmail.com>                         |                    |
```
Comment (sergeyklay) [inline phalcon/validation/validator/ip.zep]: Extra `                    |`

File: tests/unit/Validation/Validator/IpTest.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+use Phalcon\Test\Module\UnitTest;
+use Phalcon\Validation;
+use Phalcon\Validation\Validator\Ip;
+
+/**
+ * \Phalcon\Test\Unit\Validation\Validator\DateTest
+ * Tests the \Phalcon\Validation\Validator\Date component
+ *
+ * @copyright (c) 2011-2016 Phalcon Team
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]: `2011-2017`

File: tests/unit/Validation/Validator/IpTest.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            $validation->add('ip', new Ip([
+                'message' => 'This is a test message',
+                'version' => Ip::VERSION_4,
+                'allowPrivate' => true,
+                'allowReserved' => true,
+                'allowEmpty' => true,
+            ]));
+
+            $messages = $validation->validate(['ip' => '127.0.0.1']);
+            $this->assertEmpty($messages, 'Ip address should be valid');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '127.0.0.1']))->isEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate(['ip' => '192.168.10.20']);
+            $this->assertEmpty($messages, 'Ip address should be valid');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '192.168.10.20']))->isEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate(['ip' => '']);
+            $this->assertEmpty($messages, 'Empty ip address should be valid.');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '']))->isEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate(['ip' => '2001:cdba:0000:0000:0000:0000:3257:9652']);
+            $this->assertNotEmpty($messages, 'Ip version 6 should not be valid');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '2001:cdba:0000:0000:0000:0000:3257:9652']))->notEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            $validation->add('ip', new Ip([
+                'message' => 'This is a test message',
+                'version' => Ip::VERSION_4 | Ip::VERSION_6,
+                'allowPrivate' => false,
+                'allowReserved' => false,
+                'allowEmpty' => false,
+        	  ]));
+
+            $messages = $validation->validate(['ip' => '127.0.0.1']);
+            $this->assertNotEmpty($messages, 'Ip address should not be valid');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '127.0.0.1']))->notEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate(['ip' => '192.168.10.20']);
+            $this->assertNotEmpty($messages, 'Ip address should not be valid');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '192.168.10.20']))->notEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate(['ip' => '']);
+            $this->assertNotEmpty($messages, 'Empty ip address should not be valid.');
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($validation->validate(['ip' => '']))->notEmpty();
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+                    'ip' => false,
+                    'anotherIp' => true,
+                ],
+            ]));
+
+            $messages = $validation->validate([
+                'ip' => '127.0.0.1', 
+                'anotherIp' => '127.0.0.1',
+            ]);
+            expect($messages->count())->equals(1);
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($messages)->count(1);
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate([
+                'ip' => '192.168.10.20', 
+                'anotherIp' => '192.168.10.20',
+            ]);
+            expect($messages->count())->equals(1);
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($messages)->count(1);
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate([
+                'ip' => '192.168.10.20', 
+                'anotherIp' => '',
+            ]);
+            expect($messages->count())->equals(0);
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($messages)->count(0);
```

File: tests/unit/Validation/Validator/IpTest.php [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+            $messages = $validation->validate([
+                'ip' => '2001:cdba:0000:0000:0000:0000:3257:9652', 
+                'anotherIp' => '2001:cdba:0000:0000:0000:0000:3257:9652',
+            ]);
+            expect($messages->count())->equals(1);
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/IpTest.php]:
```php
expect($messages)->count(1);
```

**Review by sergeyklay — CHANGES_REQUESTED** [review]
We use the 3.0.x branch is only for bugfixes. Could you please submit this PR to the 3.1.x branch?

**Comment by Jurigag** [comment]
It's already onto 3.1.x @sergeyklay

**Comment by sergeyklay** [comment]
Oh, yes. Well, we need tune up the IpTest

**Comment by Jurigag** [comment]
@gguridi will you go back to this PR?

**Comment by Jurigag** [comment]
If this will be accepted for 3.1.x by @sergeyklay update also https://github.com/phalcon/cphalcon/blob/3.1.x/phalcon/mvc/model/validator/ip.zep to add here `@see`

**Comment by gguridi** [comment]
The CHANGELOG.md is again in conflict.

But it has a top line 3.2.0 (2017-XX-XX)... shouldn't the branch 3.1.x know nothing about 3.2.x branch? It's like the past knowing what's gonna happen in the future...

**Comment by Jurigag** [comment]
I agree

**Comment by Jurigag** [comment]
Well i guess right now you should change base to 3.2.x so we will add it in 3.2.0

**Comment by gguridi** [comment]
Changed base to 3.2.x

**Review by Jurigag — CHANGES_REQUESTED** [review]
Update also https://github.com/phalcon/cphalcon/blob/3.1.x/phalcon/mvc/model/validator/ip.zep to add here `@see`

**Comment by Jurigag** [comment]
@gguridi can you rebase and squash commits?

**Comment by niden** [comment]
@gguridi Thank you for all this work. Apologies for not getting to this earlier.

Would you be so kind as to rebase this with the current `3.4.x` branch and also issue a PR with the same work in for the `4.x` branch?

Thank you again

**Comment by gguridi** [comment]
@niden I will try to do it as soon as possible.

**Comment by niden** [comment]
@gguridi Thank you Sir!

**Comment by gguridi** [comment]
@niden Updated this PR and created [PR 13574](https://github.com/phalcon/cphalcon/pull/13574) for the 4.0.x branch. Seems that branch 3.4.x tests are broken though.

**Pull request closed without merging**